### PR TITLE
Add Naga variant to ShaderSource

### PR DESCRIPTION
**Pull request by rttad**
**Connections**
Implemention of #2788

**Description**
Adds a new `Naga(Cell<naga::Module>)` variant to ShaderSource. The module has to be wrapped in `std::cell::Cell` because `naga::Module` does not implement Copy.

**Testing**
`cargo check --features naga` passed on native target and WebAssembly.

### CODE REVIEW

File: wgpu/src/lib.rs [hunk 1]
```diff
@@ -812,6 +812,10 @@ pub enum ShaderSource<'a> {
     },
     /// WGSL module as a string slice.
     Wgsl(Cow<'a, str>),
+    /// Naga module.
+    #[cfg(feature = "naga")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "naga")))]
+    Naga(std::cell::Cell<naga::Module>),
```
Comment (cwfitzgerald) [inline wgpu/src/lib.rs:818]: Not too happy about this, but can't think of a better solution.

**Comment by jimblandy** [comment]
If we use `Cell` in `wgpu::ShaderSource`, then that makes it not implement `Sync`, which can be a pain. I think it should be fine if we simply pass in the `naga::Module` by value:

- give `ShaderSource` a variant `Naga(naga::Module)`, that owns the module,
- change `create_shader_module` to simply take the `ShaderModuleDescriptor` by value, not by reference, and
- plumb that through `wgpu::Context::device_create_shader_module`.

Then the `direct` back end can just move the `naga::Module` into the `wgpu_hal::ShaderInput` it creates.

This is a breaking API change, but now's the time to make such changes. I've pushed changes to pass `ShaderModuleDescriptor` by value here, in case you want to rebase your change on that: https://github.com/jimblandy/wgpu/tree/pass-ShaderModuleDescriptor-by-value
It's a big patch, but it's mechanical.

**Comment by rttad** [comment]
Had to force push to fix a git manipulation. The change should now be rebased correctly.

**Review by cwfitzgerald — APPROVED** [review]
Cool beans

**Pull request merged**